airbyte-integrations/connectors/source-optimizely-campaigns/source_optimizely_campaigns/source.py

```diff
@@ -0,0 +1,371 @@
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        decoded_response = response.json()
+        for link in decoded_response.get("links", []):
+            if link.get("rel", None) == "next":
+                decoded_response.update({"has_more": True, "next": link.get("href")})
+                break
+
+        if decoded_response.get("has_more", False) and decoded_response.get("elements", []):
+            parsed = urlparse.urlparse(decoded_response.get("next"))
+            return parse_qs(parsed.query)
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        # Optimizely default pagination is 100, max is 100000
+        params = {"limit": self.API_LIMIT}
+
+        # Handle pagination by inserting the next page's token in the request parameters
+        if next_page_token:
+            params.update(next_page_token)
+
+        return params
+
+    def request_headers(self, **kwargs) -> Mapping[str, Any]:
+        if self.client_id:
+            return {"Optimizely-Account": self.client_id}
+
+        return {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+
+        yield from response_json.get("elements", [])
+
+
+class IncrementalOptimizelyStream(OptimizelyStream, ABC):
+    cursor_field = "modified"
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+
+        return {self.cursor_field: max(latest_record.get(self.cursor_field, ""), current_stream_state.get(self.cursor_field, ""))}
+
+    def request_params(self, stream_state=None, **kwargs):
+        stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        params["modified[gte]"] = stream_state.get(self.cursor_field)
+        return params
+
+
+class RecipientLists(IncrementalOptimizelyStream):
+    primary_key = ["id"]
+    cursor_field = "modified"
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return "/recipientlists"
+
+
+class Recipients(IncrementalOptimizelyStream):
+    primary_key = ["id", "recipient_list"]
+    cursor_field = "modified"
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        recipient_list = stream_slice["list_id"]
```

Comment:
I'm not sure I can see where this stream_slice is coming from? Please let me know if I'm missing something!

These are generally defined by overriding the `stream_slices()` method (inherited from `Stream`, which `HttpStream` is a child of). More info on this [here](https://docs.airbyte.io/connector-development/cdk-python/stream-slices).